### DISCUSSION TITLE
Example code for a simple scenario, input List of Points --> output List of Triangles (List of 3 point array)?

### DISCUSSION THREAD

**Discussion by istinnstudio** [gwlucastrig/Tinfour#95] [Q&A] [answered]
Hi, I have only a general view about this library so before going deep I would like to know if it is possible to just use the core implementation in order to simply triangulate path2D polygons (or Lists of points) and get a list of the triangles, much like those libs do: 
https://github.com/jdiemke/delaunay-triangulator
https://github.com/orbisgis/poly2tri.java

Paths or List of points as input, list of triangles (3 point Lists) as output, those can then be converted to paths again and used as Graphics2D objects. The use case is art projects, not geographic information systems, so objectives are different, the coordinate precision remains critical though, as "double".
Is there a code example that can be provided here in order to cover this simple scenario, input List of points--> output List of Triangles ? 

Can we also tweak the result by using other data  as "constrains" additional "list of points" or polygons?
thanks!

**Comment by gwlucastrig** [comment, marked as answer]
If you have a Path2D object, extracting the points is a separate task.  But here's an example of how it would work if you had the points to start with.

There is more explanation on the Tinfour wiki article [Three Easy Iterators](https://github.com/gwlucastrig/Tinfour/wiki/Three-Easy-Iterators:-Edges,-Vertices,-and-Triangles)

```
  public static void main(String[] args) {
    // make a polygon, just as an example
    int s = 1; // length of side
    int n = 6; // number of sides
    IncrementalTin tin = new IncrementalTin(1.0);
    for(int i=0; i<n; i++){
      double theta = i*2*Math.PI/n;
      double x = s*Math.cos(theta);
      double y = s*Math.sin(theta);
      Vertex v = new Vertex(x, y, 1, i);
      tin.add(v);
    }
   
    // get the triangles
    for(SimpleTriangle triangle:  tin.triangles()){
      System.out.println(""+triangle);
    }
  }
```

**Comment by gwlucastrig** [comment]
A Delaunay triangulation always creates a convex enclosing polygon.  So if your art object is non-convex, the code example will produce some triangles that are exterior to the art object.  If this is a problem, you will have to define constraint objects, which is a somewhat more complicated problem.

To learn more about how that part of the Tinfour API works, see:  [Using Polygon Based Constraints](https://github.com/gwlucastrig/Tinfour/wiki/Tutorial-Using-Polygon-Based-Constraints)

If you need help with that, I can probably lend some assistance.

**Comment by istinnstudio** [comment]
Thank you!! very usefull reply, I have used the library with success in thousands of polygons. This example was very informative in order to quickly see the sequence of functions.
And yes, the case with some non-convex polygons is present, the external triangles effect is nice and surely usable, but I will try to find the way not to create them in those spaces. I will see if I can find the way in ... [Polygon Based Constraints](https://github.com/gwlucastrig/Tinfour/wiki/Tutorial-Using-Polygon-Based-Constraints)
Can I use the original path in order to create a Constraint for those cases?

**Comment by gwlucastrig** [comment]
The geometry from the original path may be used provided it is not self
intersecting. Also, it must be oriented counterclockwise. The Java Path2D class can
contain a few objects that are not supported  such as curved arcs.
You can use some of the techniques in the wiki article to know if a
triangle is inside your polygon or outside

**Comment by istinnstudio** [comment]
Thanks for the support, it seems that for now I cannot find a way to remove those exterior triangles. But, if there would be an example based on the basic code above (with convex and non convex source) that would be great. As a final thought, as far as I can see, there could be a method that returns boolean and if  called can "flag/mark" those triangles at the very begining of the tin creation, so all of them can be selected and used/removed afterwards without further conversions or "contains" calculations. Just thinking, not sure if it is a good idea or have application in other cases.

**Reply by micycle1** [reply]
Sounds like you might want to use `TriangleCollector.visitTrianglesConstrained()`.

**Reply by istinnstudio** [reply]
I have tried to use this "consumer" interface following the wiki examples and guessing some parts. I think I have manage to get some use out of it and print some true/false calculations with TriangleCollector.visitTrianglesConstrained().

But the problem is that when I use tin.addConstraints(constraints, true) the triangulations always dissapear. A guess is that in the current state all my polygon shapes are being converted to vertex points, without the use of any internal points (only a perimeter of 3 to n points). So I think that the "zero" triangulation  has to do with the constraint sharing the same coords as the original shape. Or some other buggy conversion part in my messy code.  So the last resort maybe is to  experiment with new morphings adding points inside polygons area,  or just "select" and remove all those exterior simple-triangles at the path2D conversion level with other spatial "contains ..." queries. So this case seems to be very specific for my shapes. I am positive, that I will get some good results, this library seems that has been developed with powerfull functions.

I am sharing some "consumer" code based on first example for future reference:

```


import java.util.ArrayList;
import org.tinfour.common.IConstraint;
import org.tinfour.common.PolygonConstraint;
import org.tinfour.common.Vertex;
import org.tinfour.standard.IncrementalTin;
import org.tinfour.utils.TriangleCollector;

/**
 *
 * @author jan
 */
public class UseOfTriangleConsumer {
   
   public static void main (String[] args){
   
      IncrementalTin tin = new IncrementalTin(1.0);
       PolygonConstraint polygonConstraint = new PolygonConstraint();     
          
ArrayList<Vertex> vertexList = new ArrayList<>();  
      int s = 1; // length of side
    int n = 1024; // number of sides
  Vertex v = null;
    for(int i=0; i<n; i++){
      double theta = i*2*Math.PI/n;
      double x = s*Math.cos(theta);
      double y = s*Math.sin(theta);
       v = new Vertex(x, y, 1, i);
      
      
vertexList.add(v);


//Declaring Array with Equal Size to the List
        Vertex[] vertexArray  = new Vertex [vertexList.size()];
      

        //Converting to Array
        for (int j = 0 ; j < vertexArray.length ; j++){
            vertexArray[i] = vertexList.get(i);
            
            
            
polygonConstraint.add(vertexArray[i]);
            
        }


    }
      


tin.add(vertexList, null); // always on top of addConstraints


ArrayList<IConstraint> constraints = new ArrayList<>();
constraints.add(polygonConstraint);
tin.addConstraints(constraints, true); // always after tin.add(vertexlist)

TriangleConsumerForSimpleTriangle consumerStr = new TriangleConsumerForSimpleTriangle();

TriangleConsumerForVertexArray consumerVa = new TriangleConsumerForVertexArray();


TriangleCollector.visitSimpleTriangles(tin, consumerStr);

TriangleCollector.visitTrianglesConstrained(tin, consumerVa);

System.out.format("Number of triangles  %4d%n", consumerStr.nTriangles);
System.out.format("Total area:          %7.2f%n", consumerStr.sumArea);

   
   System.out.format("is Ghost? ... %n", consumerStr.isGhost);
   System.out.println("is.. "+consumerStr.isGhost);
   
   
   
   
   System.out.format("VertexA Member of constrain  %n", consumerVa.isTriangleConstrainedA);
   System.out.println(" Vertex A is.. "+consumerVa.isTriangleConstrainedA);

    System.out.format("VertexB Member of constrain  %n", consumerVa.isTriangleConstrainedB);
   System.out.println(" Vertex B is.. "+consumerVa.isTriangleConstrainedB);
    System.out.format("VertexC Member of constrain  %n", consumerVa.isTriangleConstrainedC);
    System.out.println(" Vertex C is.. "+consumerVa.isTriangleConstrainedC);
   
   }

}

```


```
import java.util.function.Consumer;
import org.tinfour.common.Vertex;

public class TriangleConsumerForVertexArray implements Consumer<Vertex[]> {

	int nVertex;
      
    public  double getX_A;
    public  double getX_B;
    public  double getX_C;
    public  double getY_A;
    public  double getY_B;
    public  double getY_C;
      

    public  boolean isTriangleConstrainedA;
    public  boolean isTriangleConstrainedB;
    public  boolean isTriangleConstrainedC;


      
	@Override
	public void accept(Vertex[] va) {
	  nVertex++;
isTriangleConstrainedA = va[0].isConstraintMember();
 isTriangleConstrainedB = va[1].isConstraintMember();  
  isTriangleConstrainedC = va[2].isConstraintMember(); 
  
  
  
  getX_A = va[0].getX();
  getX_B = va[1].getX();
  getX_C = va[2].getX();
  getY_A = va[0].getY();
  getY_B = va[1].getY();
  getY_C = va[2].getY();
  
	}

}
```

```

import java.util.function.Consumer;
import org.tinfour.common.IConstraint;
import org.tinfour.common.SimpleTriangle;

public class TriangleConsumerForSimpleTriangle implements Consumer<SimpleTriangle> {

	public int nTriangles;
	public double sumArea;

     public boolean isGhost;
      
   public IConstraint  theContainingRegion;
      
	@Override
	public void accept(SimpleTriangle t) {
	  nTriangles++;
	  sumArea += t.getArea();
   
        isGhost = t.isGhost();
        
     theContainingRegion =   t.getContainingRegion();

     
	}

}
```

**Reply by gwlucastrig** [reply]
It may be that some of your code got scrambled when you pasted it to this discussion...  Perhaps some braces got dropped.  Looking at your main, the nesting seems a little off.  I suggest the code below.  Also, the "visit" methods loop through all the triangles in the Delaunay mesh, so your consumers will be called multiple times.  The print statements at the end of your main will print the results from the last triangle visited.  As a test, you could try putting print statements into your consumers.  Though, if you do, I suggest you start off with a much smaller set of vertices so you are not overwhelmed by a massive lines of text output.

As an aside, for a sufficiently large set of input vertices, the number of triangles in a Delaunay approaches 2 times the number of vertices.

```
  IncrementalTin tin = new IncrementalTin(1.0);
    PolygonConstraint polygonConstraint = new PolygonConstraint();

    ArrayList<Vertex> vertexList = new ArrayList<>();
    int s = 1; // length of side
    int n = 1024; // number of sides
    Vertex v = null;
    for (int i = 0; i < n; i++) {
      double theta = i * 2 * Math.PI / n;
      double x = s * Math.cos(theta);
      double y = s * Math.sin(theta);
      v = new Vertex(x, y, 1, i);
      vertexList.add(v);
    }

    //Declaring Array with Equal Size to the List
    Vertex[] vertexArray = vertexList.toArray(new Vertex[0]);

    //Converting to Array
    for (int j = 0; j < vertexArray.length; j++) {
        polygonConstraint.add(vertexArray[j]);
    }

    tin.add(vertexList, null); // always on top of addConstraints

    ArrayList<IConstraint> constraints = new ArrayList<>();
    constraints.add(polygonConstraint);
    tin.addConstraints(constraints, true); // always after tin.add(vertexlist)

    TriangleConsumerForSimpleTriangle consumerStr = new TriangleConsumerForSimpleTriangle();

    TriangleConsumerForVertexArray consumerVa = new TriangleConsumerForVertexArray();

    TriangleCollector.visitSimpleTriangles(tin, consumerStr);

    TriangleCollector.visitTrianglesConstrained(tin, consumerVa);
```

**Reply by istinnstudio** [reply]
Sorry for the messy code, I try to code for art n my level of coding is childish (self-taught) ,  the outcome is a combination of happy accidents and scrambled math.

Some notes on my last trial and error session, following your suggestions, reforming code etc, I have some interesting results with good visual effect. When using constrains with the same coords as the original shape,  tin.addConstraints(constraints, true); no triangulation gets activated (zero simple triangles), the output is null. 
But then I add different shapes diverted from the original, using the excellent java port of clipper-2 lib https://github.com/micycle1/Clipper2-java from "[micycle1](https://github.com/micycle1)", the triangulation is being created with new formation as long as the shape can accept a valid internal or external offset. The result is outstanding. 
I also noticed strange messages that output negative values for the area of the constrains. Maybe it has to do with direction of the paths :
```
polygonConstraint  = PolygonConstraint, area=-3.3197640553867505
constraints = [PolygonConstraint, area=-3.3197640553867505]
```

As for the external triangle that has to be removed, I have not any progress, but I am sure I will do this in the future somehow, I need time to place functions in the right logic.

This discussion helped me discover basic functionality and this lib, will remain for sure, as one of my coding wepons for the future. Thanks !!

**Reply by micycle1** [reply]
You may find this code of mine useful. It possibly doing a similar thing to what you're trying to achieve -- constrained triangulations of shapes (possibly with holes) ultimately for art:

[This](https://github.com/micycle1/PGS/blob/d2bb970cfafdb5be2a25f8cc6ed7e6fd6de60657/src/main/java/micycle/pgs/PGS_Triangulation.java#L264) method triangulates shapes and [this](https://github.com/micycle1/PGS/blob/d2bb970cfafdb5be2a25f8cc6ed7e6fd6de60657/src/main/java/micycle/pgs/PGS_Triangulation.java#L97) one outputs a collection of triangle shapes.

Note this comment:

> In Tinfour, the shape exterior (constraint) must be CCW and the holes must be CW.

**Reply by istinnstudio** [reply]
Thanks for the support and guidance , I really appreciate it, 
 [gwlucastrig](https://github.com/gwlucastrig) & [micycle1](https://github.com/micycle1)... your work is outstanding. @micycle1 I wish the "processing.org" PGS code had a chance to live in native java easily... I will take a deeper look when I am ready to understand "processing.org" idiosyncrasies. Capabilities look awsome.
I would like to share my solution to the issue of the excess triangles,  it is based on the simple triangle consumer as discussed above. 
I had tried to remove those triangles on the produced path2D but I have failed because I call "triangulation" on multiple levels on top of eatch other and the outcome was messed up. Then I have realized that it is indeed better to see the "consumer" as a convenient i/o interface and fix this on the fundamental level of the simple triangles. So this is the code of the consumer, the driver code I guess can be imagined as it is strictly dependent on my project. I create a centroid point for every simple triangle, then I check if this point is contained by the original path, if not, I remove the triangle..., to sum up, this discussion helped a lot!

```

import java.awt.geom.AffineTransform;
import java.awt.geom.Path2D;
import java.awt.geom.Point2D;
import java.util.function.Consumer;
import org.tinfour.common.IConstraint;
import org.tinfour.common.SimpleTriangle;

public class TriangleConsumerForSimpleTriangle implements Consumer<SimpleTriangle> {

	public int nTriangles;
	public double sumArea;

     public boolean isGhost;
     
     public boolean isExternalSimpleTriangle;
      
   public IConstraint  theContainingRegion;
   public Path2D path2dGetFromSimpleTriangle;

   public Path2D getPath2dGetFromSimpleTriangle() {
      return path2dGetFromSimpleTriangle;
   }

   public void setPath2dGetFromSimpleTriangle(Path2D path2dGetFromSimpleTriangle) {
      this.path2dGetFromSimpleTriangle = path2dGetFromSimpleTriangle;
   }

	@Override
	public void accept(SimpleTriangle t) {
	  nTriangles++;
	  sumArea += t.getArea();
        isGhost = t.isGhost();
     theContainingRegion =   t.getContainingRegion();

     AffineTransform defaultTransform = new AffineTransform();
     path2dGetFromSimpleTriangle = t.getPath2D(defaultTransform);
	}

   // ADD A SPATIAL QUERY: THE CENTROID OF A SIMPLE TRIANGLE, AGAINST ITS SOURCE PATH, WITH THE PATH2D.CONTAINS NATIVE METHOD, ADD THE RESULT AS BOOLEAN
    // DRIVER FOR LEVEL 1 TRIANGULATION:
      /* // get the triangles
      for (SimpleTriangle triangle : tin.triangles()) {
      // System.out.println(String.valueOf(triangle));
      
      if (!triangle.isGhost()) { //not needed, JUST IN CASE
      // FILTERING INVALID SIMPLETRIANGLES WITH BOOLEAN FROM SIMPLETRIANGLE CONSUMER
      consumerSimTri.accept(triangle, path2d);
      boolean isExt = consumerSimTri.isExternalSimpleTriangle;

      if (isExt){
      simpleTriangleList.add(triangle); //all triangles
      }
      }
      }*/
      
      /*// DRIVER FOR LEVEL 2 TRIANGULATION:
      // get the triangles
      for (SimpleTriangle triangle : tin.triangles()) {
      if (!triangle.isGhost()) { //not needed, JUST IN CASE
      
      consumerSimTri.accept(triangle, path2d);
      boolean isExt = consumerSimTri.isExternalSimpleTriangle;
      System.out.println("is ext = "+isExt);
      
      if (isExt){
      simpleTriangleList.add(triangle); //all triangles
      
      // warning
      //NEW VERSION
      Path2D simpleTinVertexPath2d = (triangle.getPath2D(null));
      triangulatedTinVertexPath2d.append(simpleTinVertexPath2d, true);
      simpleTrianglePath2DList.add(simpleTinVertexPath2d);
      
      }
      }
      
      }*/

//  NEW CONSUMER ACCEPT, IT ACTUALLY WORKS!
	public void accept(SimpleTriangle t, Path2D path) {
	  nTriangles++;
	  sumArea += t.getArea();
        isGhost = t.isGhost();
     theContainingRegion =   t.getContainingRegion();

     AffineTransform defaultTransform = new AffineTransform();
     path2dGetFromSimpleTriangle = t.getPath2D(defaultTransform);
     
            Point2D triPoA = new Point2D.Double(t.getVertexA().x, t.getVertexA().y);
            Point2D triPoB = new Point2D.Double(t.getVertexB().x, t.getVertexB().y);
            Point2D triPoC = new Point2D.Double(t.getVertexC().x, t.getVertexC().y);
     
   Point2D  centroidOfSimpleTriangle = centroidTriangle(triPoA, triPoB, triPoC);
     isExternalSimpleTriangle = path.contains(centroidOfSimpleTriangle);
     
	} 
      
        // triangle centroid mode
   public Point2D centroidTriangle(Point2D p1, Point2D p2, Point2D p3) {
      // coordinate of the vertices in param

      // Formula to calculate centroid
      double x = (p1.getX() + p2.getX() + p3.getX()) / 3;
      double y = (p1.getY() + p2.getY() + p3.getY()) / 3;

      //System.out.print(setprecision(3));
//    System.out.println("Centroid = " + "(" + x + ", " + y + ")");
      Point2D centroidTriangle = new Point2D.Double(x, y);

      return centroidTriangle;
   }
}





```

**Reply by gwlucastrig** [reply]
Thanks for the kind words.  Glad things worked out.

**Reply by micycle1** [reply]
Also, PGS does live in native Java! You can use it as a maven library alongside Processing to code in Java (as I generally do).

**Reply by istinnstudio** [reply]
> Also, PGS does live in native Java! You can use it as a maven library alongside Processing to code in Java (as I generally do).
@micycle1

I will defenately try that out, I generally avoid processing "extends PApplet" class because my project is based mainly on all java graphics2D native methods for various reasons and shapes are being formed in Path2D.Double. Although I can get some PImage "bufferedimages" from processing to work with, I have never tried to work with  i/o vector data, points, shapes etc.

So when you say : "The focus of the library is on visualisation rather than providing underlying data structures. To this end all methods in the library are static and most of them take in and return [PShapes](https://processing.org/reference/PShape.html) or [PVectors](https://processing.org/reference/PVector.html)."

you meen that if I can code some simple convertions between PShape <--> Path2D and PVector <--> Point2D (like I do for the tinfour for Vertex), then we can use the library as we would do in processing ide or extending the processing.org PApplet class? Is this possible? What about the "float" precision of the processing.org coordinates?

**Reply by micycle1** [reply]
The conversions are [already](https://github.com/micycle1/PGS/blob/7894ca270455ccf382d8d0df555ff4a9f5101511/src/main/java/micycle/pgs/PGS_Conversion.java#L1021) coded, so you needn't even work in Processing at all -- you can work with what you have and convert awt shapes, process them with PGS, then convert back to display them.

**Reply by istinnstudio** [reply]
Nice!! I am curious how the float precision of Processing Objects will "react" with double precision Path2Ds, and if this "downgrade" will affect visual appearence in high magnifications or very small areas... (I imagine hybrids that both co-exist keeping original shape boundaries double, filling them inside with pgs float elements) nevertheless, so many oportunities are being opened and this is great.

**Reply by micycle1** [reply]
The precision loss of double->float conversion is negligible (think at most ~1e-7?), particularly in the context of computer graphics.

**Reply by istinnstudio** [reply]
Issues with complex drawings with thousands of polygons do exist. But I will have to test in order to find out if those limits can be reached, e.g. in svg exports. A related post can be found here : https://blog.generativedarkroom.com/why-i-switched-from-processing-to-openrndr-b50c931d21ae
